File: tests/instr/gather/instr_gather.ml
open Optitrust
open Target

let _ = Flags.check_validity := true

let _ = Run.script_cpp (fun _ ->
  (* TODO: flag to prevent moving deps
  !! Trace.failure_expected (fun _e -> true) (fun () ->
    Instr.(gather_targets ~dest:(GatherAtLast)) [cFunBody "foo"; cVarDef ""]);
    *)

  !! Instr.(gather_targets ~dest:(GatherAtFirst)) [cFunBody "foo"; cVarDef ""];
  (* FIXME:
  !! Trace.failure_expected (fun _e -> true) (fun _ ->
      Instr.(gather_targets ~dest:(GatherAt [tAfter; cFor "k"])) [cFunBody "foo"; cVarDef ""];
     );
  !! Trace.restore_original();
  !! Instr.(gather_targets ~dest:(GatherAt [tBefore; cFor "i"])) [cFunBody "foo"; cVarDef ""];
*)

  !! Instr.(gather_targets ~dest:(GatherAtLast)) [cFunBody "with_deps"; cPlusEq ~lhs:[cOr [[cVar "a"]; [cVar "b"]]] ()];
)

File: tests/instr/gather/instr_gather.cpp
#include <optitrust.h>

void foo() {
  __pure();

  int x = 3;
  for (int i = 0; i < 10; i++)
  {
    // FIXME: #unop-inc-dec x++;
    x += 1;
  }

  int y = 1;
  for (int j = 0; j < 10; j++)
  {
    // FIXME: y++;
    y += 1;
  }
  int z = 5;
  for (int k = 0; k < 10; k++)
  {
    // FIXME: z++;
    z += 1;
  }
  int t = 2;
}

void with_deps() {
  __pure();

  int a = 0;
  int b = 0;
  int x = 0;
  int y = 0;
  int u = 0;
  int v = 0;

  a += 1;
  x += a;
  u += x;
  v += 1;
  y += b;
  b += 1;
}
